# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-надёжности-заёмщиков" data-toc-modified-id="Исследование-надёжности-заёмщиков-1">Исследование надёжности заёмщиков</a></span><ul class="toc-item"><li><span><a href="#Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию." data-toc-modified-id="Шаг-1.-Откройте-файл-с-данными-и-изучите-общую-информацию.-1.1">Шаг 1. Откройте файл с данными и изучите общую информацию.</a></span></li><li><span><a href="#1.-Вывод" data-toc-modified-id="1.-Вывод-1.2">1. Вывод</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных" data-toc-modified-id="Шаг-2.-Предобработка-данных-1.3">Шаг 2. Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#2.1.-Обработка-пропусков" data-toc-modified-id="2.1.-Обработка-пропусков-1.3.1">2.1. Обработка пропусков</a></span></li><li><span><a href="#2.1-Вывод" data-toc-modified-id="2.1-Вывод-1.3.2">2.1 Вывод</a></span></li><li><span><a href="#2.2.-Обработка-странных-значений" data-toc-modified-id="2.2.-Обработка-странных-значений-1.3.3">2.2. Обработка странных значений</a></span></li><li><span><a href="#2.2-Выводы" data-toc-modified-id="2.2-Выводы-1.3.4">2.2 Выводы</a></span></li><li><span><a href="#2.3.-Замена-типа-данных" data-toc-modified-id="2.3.-Замена-типа-данных-1.3.5">2.3. Замена типа данных</a></span></li><li><span><a href="#2.3.-Вывод" data-toc-modified-id="2.3.-Вывод-1.3.6">2.3. Вывод</a></span></li><li><span><a href="#2.4.-Обработка-дубликатов" data-toc-modified-id="2.4.-Обработка-дубликатов-1.3.7">2.4. Обработка дубликатов</a></span></li><li><span><a href="#2.4.-Вывод" data-toc-modified-id="2.4.-Вывод-1.3.8">2.4. Вывод</a></span></li><li><span><a href="#2.5.-Лемматизация" data-toc-modified-id="2.5.-Лемматизация-1.3.9">2.5. Лемматизация</a></span></li><li><span><a href="#2.5.-Вывод" data-toc-modified-id="2.5.-Вывод-1.3.10">2.5. Вывод</a></span></li><li><span><a href="#2.6.-Категоризация-данных" data-toc-modified-id="2.6.-Категоризация-данных-1.3.11">2.6. Категоризация данных</a></span></li><li><span><a href="#2.6.-Вывод" data-toc-modified-id="2.6.-Вывод-1.3.12">2.6. Вывод</a></span></li></ul></li><li><span><a href="#Шаг-3.-Ответьте-на-вопросы" data-toc-modified-id="Шаг-3.-Ответьте-на-вопросы-1.4">Шаг 3. Ответьте на вопросы</a></span><ul class="toc-item"><li><span><a href="#3.1.-Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.1.-Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-1.4.1">3.1. Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#3.1.-Вывод" data-toc-modified-id="3.1.-Вывод-1.4.2">3.1. Вывод</a></span></li><li><span><a href="#3.2.-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.2.-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-1.4.3">3.2. Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#3.2.-Вывод" data-toc-modified-id="3.2.-Вывод-1.4.4">3.2. Вывод</a></span></li><li><span><a href="#3.3.-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="3.3.-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-1.4.5">3.3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#3.3-Вывод" data-toc-modified-id="3.3-Вывод-1.4.6">3.3 Вывод</a></span></li><li><span><a href="#3.4.-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="3.4.-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-1.4.7">3.4. Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#3.4.-Вывод" data-toc-modified-id="3.4.-Вывод-1.4.8">3.4. Вывод</a></span></li></ul></li><li><span><a href="#Шаг-4.-Общий-вывод" data-toc-modified-id="Шаг-4.-Общий-вывод-1.5">Шаг 4. Общий вывод</a></span><ul class="toc-item"><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-1.5.1">Чек-лист готовности проекта</a></span></li></ul></li></ul></li></ul></div>

## Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd
from IPython.display import display
clientele = pd.read_csv('/datasets/data.csv')
display(clientele.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [2]:
print(clientele.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


Посмотрим на полученую информацию о таблице.

Таблица содержит 12 столбцов с различными типами данных, а именно:
- 2 столбца с типом данных float64 
- 5 столбцов с типом данных int64 
- 2 столбца с типом данных object 

Разберем, как называются столбцы и какую информацию они отображают:

- children — количество детей в семье
- days_employed — трудовой стаж в днях
- dob_days — возраст клиента в годах
- education — образование клиента
- education_id — идентификатор образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — доход в месяц
- purpose — цель получения кредита

Просмотрев данные по таблице, заметили следующие странности:

- В столбцах days_employed и total_income тип значений вещественные числа, это будет не совсем удобно для расчетов, поэтому вдальнейшем заменим данные на целочисленные.
- В столбцах employed и total_income отсутсвуют значения, их колличество одинаковое, возможно  данные остутствуют в одних и техже строках, проверим это предположение в дальнейшем и решим как поступить с данными пропусками.

In [3]:
clientele.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Посмотрев общую информацию по таблице выявили следующие странности в значениях:

- В столбце children мы видим значения равные -1 и 20. Значение -1 точно не может быть, 20 возможно, но маловероятно. Проверим вдальнейшем какое колличество строк принимает данное значение и примем решение что делать с этими данными.
- В столбце dob_years замечени минимально значение 0, скорее всего это ошибка, тк при заполнении анкеты на кредит, человек точно указывает свои пспортные данные.
- В стобце days_empoyed наблюдаем отричательные значения, скорее всего это ошибка в загрузке данных, избавимся от них взяв данные знчения по модулю

## 1. Вывод

В каждой строке таблицы содержится информация о человеке определенного возраста и пола с опредленным образованием, семейным положением и колличеством детей, который ранее обрщаля в банк для получения кредита на конкретную цель, предоставляя при этом данные о своей занятости, трудовом стаже и доходе в месяц. Также нам известно имел ли человек задолженость по возрату кредитов.

В процессе изучения таблицы были выявлены следующие проблемы с данными:

- отрицательные значения в столбце days_employed, хотя речь идет о трудовой стаже в днях.
- не соотвествие типа данных в столбцe days_emloyed (тип данных для столбца float64) хотя речь идет о трудовой стаже в днях, а это целое число
- разные регистры написания значений в столбце education
- столбцы employed и total_income с отсутсвующими значениями
- столбец children имеет значения равные -1 и 20, хотя речь идет о количестве детей
- столбец dob_years имеет минимально значение 0



Для того чтобы разобраться, влияет ли семейное положение и количество детей клиента на факт возврата кредита в срок наиболее важные столбцы children, family_status_id, debt и total_income. 

Вертнуться к [**Содержанию**](#Содержание)

## Шаг 2. Предобработка данных

### 2.1. Обработка пропусков

Проверим данные на наличие пропусков вызовом набора методов для суммирования пропущенных значений.

In [4]:
print(clientele.isnull().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


Число пропусков в столбцах days_employed  и total_income одинаково, по 2174 пропуска. Из таблицы видно, что пропуски в столбце days_employed и total_income совпадают. Значение в столбце days_employed не является для нас важными, заменим все пропущенные значения 0. Значение в столбце total_income является важным, тк один из впоросов иследования зависимость между уровнем дохода и возвратом кредита в срок. 

In [5]:
clientele[clientele['days_employed'].isnull()].head(5)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [6]:
clientele[clientele['days_employed'].isnull()].tail(5)

children  days_employed  dob_years education  education_id  \
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21489   женат / замужем                 0      M   компаньон     0   
21495  гражданский брак                 1      F   сотрудник     0   
21497   женат / замужем                 0      F   компаньон     0   
21502   женат / замужем                 0      F   сотрудник     0   
21510   женат / замужем                 0      F   сотрудник     0   

       total_income                           purpose  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля

In [7]:
clientele[clientele['total_income'].isnull()].head(5)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [8]:
clientele[clientele['total_income'].isnull()].tail(5)

children  days_employed  dob_years education  education_id  \
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21489   женат / замужем                 0      M   компаньон     0   
21495  гражданский брак                 1      F   сотрудник     0   
21497   женат / замужем                 0      F   компаньон     0   
21502   женат / замужем                 0      F   сотрудник     0   
21510   женат / замужем                 0      F   сотрудник     0   

       total_income                           purpose  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля

Сравним фрагменты таблиц между собой и можно сделать вывод что значения в данных столбцах отсутсуют в одинаковых строках. Это конечно же может быть ошибка заполнения значений, но и возможно что данные люди просто не предоставили данные о трудовом стаже и месячном доходе. Возможно данне люди являются или являлись зарплатными клиентами данного банка и информация об их трудовом стаже и месячном доходе находится в другой базе или им для получения кредита не нужно было подтверждать свою платежеспособность, тк у них было предодобренное предложение по кредиту. Больше склоняюсь к данной версии, тк зависимости значений в данных столбцах от других я не увидел.

Значение в столбце days_employed не является для нас важными, заменим все пропущенные значения 0 и сразу проверим пропуски в столбце days_employed

In [9]:
clientele['days_employed'].fillna(0, inplace=True)
clientele['days_employed'].isnull().sum()

0

Занчение в столбце total_income является для нас важным, и если мы заменим все значения на 0 то потеряем более 10% данных (2174/21525*100%), поэтому проверим столбец на максимальное и минимальное значение, если разброс значений будет не большим, то найдем среднее значение по столбцу и заменим все пустые значения на среднее значение, если же разброс будет большим то найдем медиана по данному столбцу и заменим все пропуски медианой. 

In [10]:
print(clientele['total_income'].max())
print(clientele['total_income'].min())

2265604.028722744
20667.26379327158


Разброс достаточно большой, найдем медиану по стоблцу и заменим на ее значение все пропуски в данном стобце и проверим колличество пропусков столбце total_income

In [11]:
income_median = clientele['total_income'].median()
print(income_median)
clientele['total_income'].fillna(income_median, inplace=True)
clientele['total_income'].isnull().sum()

145017.93753253992


0

### 2.1 Вывод

Пропуски в значения были обнаружены в двух столбцац:

- days_employed (2174 пропуска)
- total_income (2174 пропуска)

Пропуски в столбцах были пропущеный в одинаковых строках, значит это либо ошибка в выгрузке данных либо клиент не предоставил данные о его трудовом стаже и доходе в месяц. Скорее всего это можно обьяснить двумя причинами:
- он являлся и является зарплатным клиентом данного банка (данные и трудовом стаже и месчном доходе скорее всего хранятся в других базах)
- он имеет предодобренное предложение от банка(тк ранее пользовался услугами банка и показал свою состоятельность или как раз был зарплатным клиентом), которое не обязует его предоставлять данные сведенья

Исходя из это и главных вопросов которые стоят перед нами в данном анализе, мы заменяем все значения в столбце days_empoyed нулевым значением, а в столбце total_income на медиану по данному столбцу.

Вертнуться к [**Содержанию**](#Содержание)

### 2.2. Обработка странных значений

В первоначальной таблице, сразу заметили что в стобце education были предоставлены одни и теже значения, но в разных регистрах. Первоначально исправим значения в данном столбце, путем приведения всех значений в один регистр. 
Проверим колличетсво строк в столбце children со значение -1 и 20. В зависимоти от колличества таких  строк исправим данные значения.
Проверим колличество строк в столбце dob_years c 0 значение. Исправим данное значение.

Проверяем количество уникальных значений для стобца educaiton

In [12]:
clientele['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Как и было замечено ранее в данном столбце присутсвуют одинаковые данные, но в разных регистрах. Сменим регистр у данных значений на нижний и проверим колличетсво значений после изменения.

In [13]:
clientele['education'] = clientele['education'].str.lower()
clientele['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Проверим количество строк в столбце children равных -1 и 20

In [14]:
clientele['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В таблице имются 47 строк с количеством детей -1 и 76 строк с количеством детей 20. Предполагаю что дети со значением -1 были загружены с неправильным знаком, поменяем их колличество на 1, на иследование это повлияет не сильно. Предполагаю что дети со значением 20 были загружены верно, хотя и мало вероятно. Исходя из нашей таблицы таких значений 0.3%, а по статистике в России семей с 5 и более детьми около 2% поэтому данное значение трогать не будем, в дальнейшем категоризируем данные значения в многодетные семьи.

In [15]:
clientele['children'] = clientele['children'].replace(-1, 1)
clientele['children'].value_counts()

0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64

Проверяем 0ые значения в столбце dob_years

In [16]:
clientele[clientele['dob_years'] == 0].count()[0]


101

Проверрим коллиичество людей в возрасте меньше или равным 18 лет

In [17]:
clientele[clientele['dob_years'] <= 18].count()[0]

101

Колличетство людей с данными значениями совпадает, значит это одни и теже люди. Так кредит в России обычно можно получить с 18 лет, заменим все 0 на 18. И проверим выполнилось ли данное действие.

In [18]:
clientele['dob_years'] = clientele['dob_years'].replace(0, 18)
clientele[clientele['dob_years'] == 0].count()[0]

0

Приведем все отрицательные значения в столбце days_employed к положительным методом abs() и проверим замену

In [19]:
clientele['days_employed'] = clientele['days_employed'].abs()
clientele['days_employed'].describe()

count     21525.000000
mean      60156.419005
std      133355.929525
min           0.000000
25%         610.652074
50%        1808.053434
75%        4779.587738
max      401755.400475
Name: days_employed, dtype: float64

### 2.2 Выводы

В стобце education привели все значения в один регистр. 
В столбце children заменили значения -1 на 1 и 20 оставили без изменений
В столбце dob_years выяснили что люди с 0ым возрастом это люди в возрасте меньше или равным 18 лет. Тк кредит можно получать с 18 лет заменили все значения 0 на 18. Поменяли все отрицательные значения в столбце days_employed.

Вертнуться к [**Содержанию**](#Содержание)

### 2.3. Замена типа данных

В первоначальной таблице заметили, что есть не соотвествие типа данных в столбцe 'days_employed' (тип данных для столбца float64) хотя речь идет о трудовой стаже в днях, а это целое число, изменим тип данных для данного стобца. Так же считаю необходимым заменить тип данных в столбце total_income на целосичленный, тк врядли значение дохода в месяц с копейками врядли координально повлияют на анализ, в тоже время без этих значений мы сможем иметь вид значений более приятных для чтения и программа затратит на это меньше времени.

Меняем вещественные значения в столбце days_employed на целочисленные

In [20]:
clientele['days_employed'] = clientele['days_employed'].astype('int')

Меняем вещественные значения в столбце total_income на целочисленные

In [21]:
clientele['total_income'] = clientele['total_income'].astype('int')

### 2.3. Вывод

Заметив не соотвествие типов значений в первоначальной таблице по столбцу days_empoyed заменили их на int64. Так же для наглядности и упрощения исчислений по стобцу total_income тоже заменили тип данных на int64. 

### 2.4. Обработка дубликатов

In [22]:
clientele.duplicated().sum()

71

In [23]:
clientele.loc[clientele.duplicated()]


children  days_employed  dob_years education  education_id  \
2849          0              0         41   среднее             1   
3290          0              0         58   среднее             1   
4182          1              0         34    высшее             0   
4851          0              0         60   среднее             1   
5557          0              0         58   среднее             1   
...         ...            ...        ...       ...           ...   
20702         0              0         64   среднее             1   
21032         0              0         60   среднее             1   
21132         0              0         47   среднее             1   
21281         1              0         30    высшее             0   
21415         0              0         54   среднее             1   

          family_status  family_status_id gender income_type  debt  \
2849    женат / замужем                 0      F   сотрудник     0   
3290   гражданский брак                 1      F   пенсионер     0   
4182   гражданский брак                 1      F   сотрудник     0   
4851   гражданский брак                 1      F   пенсионер     0   
5557   гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
20702   женат / замужем                 0      F   пенсионер     0   
21032   женат / замужем                 0      F   пенсионер     0   
21132   женат / замужем                 0      F   сотрудник     0   
21281   женат / замужем                 0      F   сотрудник     0   
21415   женат / замужем                 0      F   пенсионер     0   

       total_income                            purpose  
2849         145017            покупка жилья для семьи  
3290         145017                    сыграть свадьбу  
4182         145017                            свадьба  
4851         145017                            свадьба  
5557         145017                    сыграть свадьбу  
...             ...                                ...  
20702        145017         дополнительное образование  
21032        145017              заняться образованием  
21132        145017                       ремонт жилью  
21281        145017  покупка коммерческой недвижимости  
21415        145017                  операции с жильем  

[71 rows x 12 columns]

Мы видим что в таблице имеется 71 дубликат. Дубликаты образовались из за того что в первоначальных данных отсутствовали значения в столбцах days_employed и total_income и нам пришлось заменить их. Это вполне могут быть разные люди, тк в выборке более 21000 значений а в столбцах family_status, gender, income_type, purpose разброс значений от 3 до 30, вполне вероятно что поле того как их трудовой стаж и месячный доход стали равны, значения в каждом столбце отдельной строки стали совпадать. Удалим все дубликаты и проверим выполнение.

In [24]:
clientele = clientele.drop_duplicates()
clientele.duplicated().sum()

0

### 2.4. Вывод

В первоначальной таблице были замечены одинаковые значения в стобце education, но записанные в разных регистрах. Привели значения из данного столбца к единому регистру. Проверили все остальные столцы на такуюже ошибку в значениях. Далее проверили наличие дубликатов во всей таблице, их было обнаружено 71. Причина появления данных дубликатов заключается в том, что в первоначальных данных отсутствовали значения в столбцах days_employed и total_income. 

Вертнуться к [**Содержанию**](#Содержание)

### 2.5. Лемматизация

Для того чтобы на вопрос исследования как разные цели кредита влияют на его возврат в срок, необходимо провести лематизаия по столбцу purpose. В данном столбце есть схожие значения, но написанные разными формами слов. Проведем лемматизаию данного столбца  и определим какие слова встречаются наиболее часто.

Создадим переменную clientele_purpose содержащую в себе все уникальные значения столбца purpose.

In [25]:
clientele_purpose = str(clientele['purpose'].value_counts().index.tolist())
clientele_purpose

"['свадьба', 'на проведение свадьбы', 'сыграть свадьбу', 'операции с недвижимостью', 'покупка коммерческой недвижимости', 'операции с жильем', 'покупка жилья для сдачи', 'операции с коммерческой недвижимостью', 'жилье', 'покупка жилья', 'покупка жилья для семьи', 'строительство собственной недвижимости', 'недвижимость', 'операции со своей недвижимостью', 'строительство жилой недвижимости', 'покупка недвижимости', 'покупка своего жилья', 'строительство недвижимости', 'ремонт жилью', 'покупка жилой недвижимости', 'на покупку своего автомобиля', 'заняться высшим образованием', 'автомобиль', 'сделка с подержанным автомобилем', 'свой автомобиль', 'автомобили', 'на покупку подержанного автомобиля', 'на покупку автомобиля', 'приобретение автомобиля', 'дополнительное образование', 'сделка с автомобилем', 'высшее образование', 'образование', 'получение дополнительного образования', 'получение образования', 'профильное образование', 'получение высшего образования', 'заняться образованием']"

Проведем лемматизацию переменной clientele_purpose методом pymystem3 для определения наиболее часто встречаемых слов

In [26]:
from pymystem3 import Mystem
m = Mystem()
lemmas = m.lemmatize(clientele_purpose)

from collections import Counter
print(Counter(lemmas))

Counter({' ': 59, "', '": 37, 'недвижимость': 10, 'покупка': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'на': 4, 'операция': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'высокий': 3, 'получение': 3, 'коммерческий': 2, 'для': 2, 'жилой': 2, 'заниматься': 2, 'сделка': 2, 'дополнительный': 2, "['": 1, 'проведение': 1, 'сыграть': 1, 'сдача': 1, 'семья': 1, 'собственный': 1, 'со': 1, 'ремонт': 1, 'подержанный': 1, 'подержать': 1, 'приобретение': 1, 'профильный': 1, "']\n": 1})


Выделим 5 обобщеных целей получения кредита:

- Жилье - цели связанные с приобретение/строительством/ремонтом жилья 
- Автомобиль - цели связанные с покупка собственного/поддержанного автомобиля 
- Образование - цели связанные с получение первоначального/дополнительного/профильного образования
- Свадьба - цели связанные со свадьбой 
- Недвижимость - цели связанные с приобритением комерческой/собственной/жилой недвижимости



### 2.5. Вывод

Лематизация столбца purpose помогла определить 5 обобщеных целей для получения кредита, это - жилье, автомобиль, образование, свадьба и недвижимость. В следующем этапе проведем категоризацию по данным целям, чтобы ответить на  вопрос данного иследдования. 

Вертнуться к [**Содержанию**](#Содержание)

### 2.6. Категоризация данных

Проведем категоризацию значений по следующим столбцам:
- total_income (разобъем на классы: бедный, средний, состоятельный и богатый)
- children (по колличеству детей: нет детей, один ребенок, два ребенка, многодетный)
- purpose (обощенные цели кредита: жилье, автомобиль, образование, свадьба и недвижимость)

Для этого напишем функции которые построчно пробегают по всей таблице ищут ключевые значения, и в зависимости от них назначают нужную категорию

In [27]:
def total_income_group(row):
    if row['total_income'] <= 36000 :
        return 'бедный'
    if row['total_income'] <= 100000 :
        return 'средний'
    if row['total_income'] < 500000 :
        return 'состоятельный'
    return 'богатый'

def children_group(row):
    if row['children'] == 0:
        return 'нет детей'
    if row['children'] == 1:
        return '1 ребенок'
    if row['children'] == 2:
        return '2 ребенка'
    return 'многодетные'

def purpose_group(row):
    purpose_lemmas = m.lemmatize(row['purpose'])
    if 'жилье' in purpose_lemmas:
        return 'жилье'
    elif 'автомобиль' in purpose_lemmas:
        return 'автомобиль'
    elif 'образование' in purpose_lemmas:
        return 'образование'
    elif 'свадьба' in purpose_lemmas:
        return 'свадьба'
    elif 'недвижимость' in purpose_lemmas:
        return 'недвижимость'         

Добавим в основную таблиц три столбца:
- total_income_group с категориями по доходу в месяц
- children_group с категориями по количеству детей
- purpose_group с категориями по целям кредита

In [28]:
clientele['total_income_group'] = clientele.apply(total_income_group, axis=1)
clientele['children_group'] = clientele.apply(children_group, axis=1)
clientele['purpose_group'] = clientele.apply(purpose_group, axis=1)

Проверим добавление столбцов

In [29]:
clientele.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_group children_group purpose_group  
0               покупка жилья      состоятельный      1 ребенок         жилье  
1     приобретение автомобиля      состоятельный      1 ребенок    автомобиль  
2               покупка жилья      состоятельный      нет детей         жилье  
3  дополнительное образование      состоятельный    многодетные   образование  
4             сыграть свадьбу      состоятельный      нет детей       свадьба

### 2.6. Вывод

Провели категоризацию значений по следующим столбцам:
- total_income 
- children 
- purpose
Категоризовав все необходимые данные в таблице получили необходимый вид для того чтобы ответить на вопросы иследования.

Вертнуться к [**Содержанию**](#Содержание)

## Шаг 3. Ответьте на вопросы

Поменяем для наглядности значения в столбце debt c 0 и 1 на 'no' и 'yes' соотвественно.

In [30]:
clientele['debt'] =clientele['debt'].replace(0, 'no')
clientele['debt'] =clientele['debt'].replace(1, 'yes')

### 3.1. Есть ли зависимость между наличием детей и возвратом кредита в срок?

Сделаем сводную таблицу children_pivot чтобы понять зависимость возрата кредита от количества детей

In [31]:
children_pivot = clientele.pivot_table(index=['children_group'], columns='debt', values='children', aggfunc='count')
children_pivot['children_debt'] = children_pivot['yes']/children_pivot['no']
display( children_pivot.style.format({'children_group': "{:}",'children_debt': "{:.2%}"}) )


### 3.1. Вывод

Заемщики у которых нет детей, меньше других нарушают кредитные обязательства. 

### 3.2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

Сделаем сводную таблицу family_pivot чтобы понять зависимость возрата кредита от семейного статуса

In [32]:
family_pivot = clientele.pivot_table(index=['family_status'], columns='debt', values='family_status_id', aggfunc='count')
family_pivot['family_debt'] = family_pivot['yes']/family_pivot['no']
display( family_pivot.style.format({'family_status': "{:}",'family_debt': "{:.2%}"}) )


### 3.2. Вывод

Заемщики с не узаконеными отношениями, чаще других нарушают кредитные обязательства.  

### 3.3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Сделаем сводную таблицу total_income_pivot чтобы понять зависимость возрата кредита от уровня дохода в месяц

In [33]:
total_income_pivot = clientele.pivot_table(index=['total_income_group'], columns='debt', values='total_income', aggfunc='count')
total_income_pivot['total_income_debt'] = total_income_pivot['yes']/total_income_pivot['no']
display( total_income_pivot.style.format({'total_income_group': "{:}",'total_income_debt': "{:.2%}"}) )

### 3.3 Вывод

Заемщики с уровнем дохода <= 36000 рублей, меньше других нарушают кредитные обязательства.

### 3.4. Как разные цели кредита влияют на его возврат в срок?

Сделаем сводную таблицу purpose_pivot чтобы понять зависимость возрата кредита от целей кредита

In [34]:
purpose_pivot = clientele.pivot_table(index=['purpose_group'], columns='debt', values='purpose', aggfunc='count')
purpose_pivot['purpose_debt'] = purpose_pivot['yes']/purpose_pivot['no']
display( purpose_pivot.style.format({'purpose_group': "{:}",'purpose_debt': "{:.2%}"}) )

### 3.4. Вывод

Заемщики берущие кредит на цели связанные с приобретение/строительством/ремонтом жилья, меньше других нарушают кредитные обязательства.

Вертнуться к [**Содержанию**](#Содержание)

## Шаг 4. Общий вывод

Отвечая на поставленный вопрос "Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок", могу дать следующий ответ - семеное положение и количество детей влияет на факт погашения кредита в строк:
-заемщики с официально оформленными отношениями (или которые в прошлом были в официальном в браке) и не имеющие детей - самые ответственные заемщики;
-заемщики, состоящие в неофициальном браке или находящиеся без отношений, при этом имеющие 1 или 2 детей - самые менее ответственные заемщики.

Проаналиировав все данные, отвечачем на вопрос иследования: "Влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок?"

- заемщкики, которые состоят в официальных отношениях или состояли ранее (в разводе, вдовец/вдова) и не имеющие детей наиболее ответственные заемщики.
- заемщики, которые не состоят в официальном браке или не имеющие отношений вовсе и имеющие 1-2 ребенка наименее ответсвенные заемщики.

С учетом добавочных фактов (уровень дохода и цель вязтия кредита) выделим два типа людей:

- ответственные заемщики (нет детей, состоит/состоял в официальном браке, уровень дохода <= 36000 рублей, цели связанные с приобретение/строительством/ремонтом жилья)
- не ответственные заемщики (1-2 ребенка, не состоит в официальном браке/нет отношений, уровень дохода от 100000 до 500000 рублей, цели связанные с покупкой собственного/поддержанного автомобиля или с получение первоначального/дополнительного/профильного образования)

Вертнуться к [**Содержанию**](#Содержание)

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.